Enable downcall & upcall test suites for Java 17/18/19

ChengJin01 opened the pull request:
The changes include all test suites intended for downcall
and upcall in Java17, 18 and 19 which can be enabled
on the following platforms:
1)downcall: all supported platforms
2)upcall: AIX/PPC and Linux/PPC64LE

Signed-off-by: Cheng Jin <jincheng@ca.ibm.com>

tajila commented:
@llxia Please review these changes

llxia reviewed (approved):
test changes look good to me

tajila commented:
jenkins test sanity amac jdk17

ChengJin01 commented:
The test failures on `MacOS/Aarch64` at https://openj9-jenkins.osuosl.org/job/Test_openjdk17_j9_sanity.functional_aarch64_mac_Personal/60/tapResults/
e.g.
```
FAILED: test_vprintfFromDefaultLibWithVaList
java.util.NoSuchElementException: No value present
	at java.base/java.util.Optional.get(Optional.java:143)
	at org.openj9.test.jep389.valist.VaListTests.test_vprintfFromDefaultLibWithVaList(VaListTests.java:125)
```
was caused by the default library loading issue already detected on macOS in https://github.com/eclipse-openj9/openj9/issues/14487 and was resolved with our own fix in OpenJDK at https://github.com/ibmruntimes/openj9-openjdk-jdk17/pull/112/files#diff-9658a8d1507ae5132a614d4c38ed6589ebab5bed19c3d03f726585aeb49dce29, https://github.com/ibmruntimes/openj9-openjdk-jdk18/pull/57/files#diff-9658a8d1507ae5132a614d4c38ed6589ebab5bed19c3d03f726585aeb49dce29, https://github.com/ibmruntimes/openj9-openjdk-jdk19/pull/5/files#diff-9658a8d1507ae5132a614d4c38ed6589ebab5bed19c3d03f726585aeb49dce29 and https://github.com/ibmruntimes/openj9-openjdk-jdk/pull/457/files#diff-9658a8d1507ae5132a614d4c38ed6589ebab5bed19c3d03f726585aeb49dce29.

In addition, the tests on JDK19 are invalid given there is no our code in OpenJDK (the code was totally removed in JDKnext due to the latest APIs in Java19) and all OpenJ9 specific code has not yet merged. So I'll disable all test suites on all supported platforms in `playlist.xml` for the moment in case any partially merged code screwed them up and re-enable them all after all code listed in https://github.com/eclipse-openj9/openj9/issues/15068 are merged if this PR needs to be merged individually.

tajila commented:
jenkins test sanity amac jdk17

ChengJin01 commented:
I just updated the test cases, added a couple of tests requested by JIT (working on thunk code on zLinux). I also removed all `Valist` specific test cases as they are outdated at this point and will be totally replaced with new tests once we have `Valist` implemented on Power in OpenJDK.

knn-k commented on runtime/tests/clinkerffi/CMakeLists.txt:
```diff
@@ -23,6 +23,8 @@
 j9vm_add_library(clinkerffitests SHARED
 	downcall.c
+	upcall.c
+	valist.c
```
There is no file named `valist.c` in this directory, isn't it?

ChengJin01 replied:
Yes. The test cases for `VaList` are not ready and will be added once we have `VaList` support on Power & zLinux which I have been working on. So I just removed `valist.c` for the moment.

knn-k commented on test/functional/Java19andUp/build.xml:
```diff
@@ -49,22 +48,16 @@ SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-excepti
+				<pathelement location="${LIB_DIR}/asm.jar" />
+				<pathelement location="${build}" />
+			</classpath>
+		</javac>
```
There is a merge conflict here against PR #15767.

ChengJin01 replied:
Resolved the conflict in the building script.

ChengJin01 commented:
I just updated part of tests in `UpcallMHWithStructTests.java` for JDK17/18/19+ to support struct specific upcall on `Linux/s390x` given the padding in these tests is not required on `Linux/s390x` as explained in the PR intended for thunk generation code at https://github.com/eclipse-openj9/openj9/pull/15596#issuecomment-1204682964.

ChengJin01 commented:
I just updated this PR to add the `VaList` specific test cases in JDK19+ which is intended for the `VaList` changes at https://github.com/ibmruntimes/openj9-openjdk-jdk/pull/488 and https://github.com/ibmruntimes/openj9-openjdk-jdk19/pull/27.

0xdaryl commented on test/functional/Java17andUp/src_170/org/openj9/test/jep389/upcall/UpcallMHWithMixedSigStruTests.java:
```diff
@@ -0,0 +1,934 @@
+	}
+
+	@Test
+	public void test_addLongAndLong2FloatsFromStructByUpcallMH() throws Throwable {
```
This test fails on Windows for both HotSpot and OpenJ9.  However, the problem seems to be on the "downcall" part of the test as the elements of the struct don't even reach the C function intact (i.e., before we can even do an upcall).

For HotSpot, printing the args reaching the C function (`addLongAndLong2FloatsFromStructByUpcallMH`) before the upcall:
```
IN C: addLongAndLong2FloatsFromStruct : arg1=555555555555, arg2.elem1=388697201, arg2.elem2=0.000000, arg2.elem3=11.250000
```
For OpenJ9, printing the args reaching the C function before the upcall:
```
IN C: addLongAndLong2FloatsFromStruct : arg1=555555555555, arg2.elem1=388697201, arg2.elem2=0.000000, arg2.elem3=11.250000
```

ChengJin01 replied:
@0xdaryl, could you help double-check `values[1]` or `ffiArgs[1]` for the struct the code of `inlInternalDowncallHandlerInvokeNative` in downcall at https://github.com/eclipse-openj9/openj9/blob/b9b733900ef25ba25435e8dd557a10f4e39a4fd6/runtime/vm/BytecodeInterpreter.hpp#L5103.  I'd like to know whether or not the passed-in value is correct given the test passes on other platforms.

If the passed-in struct is correct but it gets corrupted in the C native functions, then problem should originate from `libffi` on Windows in which case we can disable this test for the moment and will need to create an issue at https://github.com/libffi/libffi/issues or https://sourceware.org/mailman/listinfo/libffi-discuss to request them to fix it up.

ChengJin01 replied:
I suspect the problem was caused by the wrong element type defined for `elem1` in `stru_Long_2_Floats` at `downcall.h`
```
typedef struct stru_Long_2_Floats {
	long elem1;  <---------
	float elem2;
	float elem3;
} stru_Long_2_Floats;
```
in which `long` is 4 bytes on Windows, which should be replaced with `LONG` (int64_t). I will update this to verify again on Windows to see how it goes.

ChengJin01 replied:
I just verified with the updated definition of `stru_Long_2_Floats` in `/runtime/tests/clinkerffi/downcall.h`
```
typedef struct stru_Long_2_Floats {
	LONG elem1;  <-------
	float elem2;
	float elem3;
} stru_Long_2_Floats;
```
and the test `test_addLongAndLong2FloatsFromStructByUpcallMH` passed on both OpenJ9 and Hotspot:
```
$ java -version
openjdk version "17.0.5-internal" 2022-10-18
OpenJDK Runtime Environment (build 17.0.5-internal+0-adhoc.Administrator.openj9-openjdk-jdk17)
Eclipse OpenJ9 VM (build jep389_419_ffi_upcall_v4_struct_v2_outofline_mh_v9_fin_v8_5-c3c03e474,
 JRE 17 Windows 10 amd64-64-Bit Compressed References 20220920_000000 (JIT enabled, AOT enabled
)
OpenJ9   - c3c03e474
OMR      - b1be89cf5
JCL      - e988b7def79 based on jdk-17.0.5+5)

$ java -version
openjdk version "17.0.2" 2022-01-18
OpenJDK Runtime Environment (build 17.0.2+8-86)
OpenJDK 64-Bit Server VM (build 17.0.2+8-86, mixed mode, sharing)
...
PASSED: test_addLongAndLong2FloatsFromStructByUpcallMH
```

ChengJin01 replied:
@0xdaryl,  please double-check the test on your side (note that the test library `clinkerffitests.dll` needs to be regenerated with the latest `downcall.h` in this PR)

0xdaryl commented on test/functional/Java19andUp/src/org/openj9/test/jep424/upcall/UpcallMHWithStructTests.java:
```diff
@@ -0,0 +1,2906 @@
+public class UpcallMHWithStructTests {
+	private static String osName = System.getProperty("os.name").toLowerCase();
+	private static String arch = System.getProperty("os.arch").toLowerCase();
+	private static boolean isLinuxS390x = osName.contains("linux") && arch.equals("s390x");
```
Windows does not require the structure padding either, so all places that check for `isLinuxS390x` will have to check for Windows as well.

I wonder if it would be better to introduce a `requiresStructurePadding` boolean here that is initialized to false for Z and Windows rather than adding OS checks for many of the tests.  It seems a bit cleaner in my opinion.

Same for all other versions of UpcallMHWithStructTests.java.

ChengJin01 replied:
I will update the test on Java17/18/19 based against the suggestion above.

tajila commented:
@ChengJin01 is testing disabled for x86?

ChengJin01 commented:
> @ChengJin01 is testing disabled for x86?

Yes. All FFI related test suites are disabled on x86 in JDK17/18/19+ for the moment
https://github.com/eclipse-openj9/openj9/blob/5b9e56396c9e0ac2bc2f2a35884cb29558f1a8d6/test/functional/Java17andUp/playlist.xml
https://github.com/eclipse-openj9/openj9/blob/5b9e56396c9e0ac2bc2f2a35884cb29558f1a8d6/test/functional/Java18andUp/playlist.xml
https://github.com/eclipse-openj9/openj9/blob/5b9e56396c9e0ac2bc2f2a35884cb29558f1a8d6/test/functional/Java19andUp/playlist.xml
given the thunk generation code on x86 at https://github.com/eclipse-openj9/openj9/pull/15926 has not yet merged (still under code review).

tajila commented:
jenkins test sanity,extended plinux,zlinux,aix,amac jdk19

ChengJin01 commented:
The downcall & upcall related test suites are enabled in JDK17/18/19+ on ppc, s390x and Aarch64 and VaList tests are only enabled on ppc in JDK19+.

tajila commented:
jenkins test sanity,extended plinux,zlinux,aix,amac jdk19

ChengJin01 commented:
I just updated that a couple of latest native test code for `VaList` which were missing in this PR.

tajila commented:
jenkins test sanity,extended plinux jdk19

tajila commented:
jenkins test sanity,extended zlinux,aix,amac jdk19
